taichi/analysis/same_statements.cpp

```diff
@@ -97,22 +122,65 @@ class IRNodeComparator : public IRVisitor {
       same = false;
       return;
     }
-
-    // If two identical statements can have different values, return false.
-    if (check_same_value_ && !stmt->is_container_statement() &&
-        !stmt->common_statement_eliminable()) {
-      same = false;
+    auto other = other_node->as<Stmt>();
+    if (stmt == other) {
       return;
     }
+
+    // If two identical statements can have different values, return false.
+    // TODO: actually the condition should be "can stmt be an operand of
+    //  another statement?"
+    const bool stmt_has_value = !stmt->is_container_statement();
+    // TODO: two identical GlobalPtrStmts cannot have different values,
+    //  but GlobalPtrStmt::common_statement_eliminable() is false.
+    const bool identical_stmts_can_have_different_value =
+        stmt_has_value && !stmt->common_statement_eliminable() &&
+        !stmt->is<GlobalPtrStmt>();
     // Note that we do not need to test !stmt2->common_statement_eliminable()
     // because if this condition does not hold,
-    // same_statements(stmt1, stmt2) returns false anyway.
+    // same_value(stmt1, stmt2) returns false anyway.
+    if (check_same_value_ && identical_stmts_can_have_different_value) {
+      if (all_states_can_be_modified_) {
+        same = false;
+        return;
+      } else {
+        bool same_value = false;
+        if (auto global_load = stmt->cast<GlobalLoadStmt>()) {
+          if (auto global_ptr = global_load->ptr->cast<GlobalPtrStmt>()) {
+            TI_ASSERT(global_ptr->width() == 1);
+            if (possibly_modified_states_.count(ir_bank_->get_async_state(
+                    global_ptr->snodes[0], AsyncState::Type::value)) == 0) {
```

Comment:
IIUC, previous we were too conservative, and returned "*not the same*" even for snodes that are only read in the tasks?

Reply:
Yes, exactly.